src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryAdder.java

```diff
@@ -133,8 +134,13 @@
                 if (cfg instanceof LibraryResolver.ResolvedLibraryConfiguration) {
                     source = ((LibraryResolver.ResolvedLibraryConfiguration) cfg).getSource();
                 }
-                librariesAdded.put(name, new LibraryRecord(name, version, kindTrusted, changelog, cfg.getCachingConfiguration(), source));
-                retrievers.put(name, cfg.getRetriever());
+                LibraryRetriever retriever = cfg.getRetriever();
+                String libraryPath = null;
+                if (retriever instanceof SCMBasedRetriever) {
+                    libraryPath = ((SCMBasedRetriever) retriever).getLibraryPath();
+                }
+                librariesAdded.put(name, new LibraryRecord(name, version, kindTrusted, changelog, cfg.getCachingConfiguration(), source, libraryPath));
```

Comment:
I don't remember exactly how all the retrieval works, so I am wondering if this could potentially cause unnecessary retrieval to happen? E.g. One git repository containing five libraries would only be retrieved one time for each scm url ignoring libraryPath, and now with this change it will retrieve/cache one copy for each libraryPath? And that might not be desired?

Reply:
Not as far as I can tell. The (SCM based) libraries are checked out:
- in the jobs own workspace
- only if needed (i.e. either caching not enabled, expired, or something else changed)

```sh
bash-4.4$ find -type d -name ".git" 
./workspace/test-pl3@libs/cd304ec6497d6f157eab60a432dcc6ecc1350d5f15d1fae9474b01115c211903/.git
./workspace/test-pl2@libs/cd304ec6497d6f157eab60a432dcc6ecc1350d5f15d1fae9474b01115c211903/.git
./workspace/test-pl@libs/cd304ec6497d6f157eab60a432dcc6ecc1350d5f15d1fae9474b01115c211903/.git
```

Library caching happens after checkout and the files land in a different directory (based on `directoryName`) and without the `.git` extras. Below two cached directories representing the vars under:

- https://github.com/sboardwell/demos/tree/main (root)
- https://github.com/sboardwell/demos/tree/main/build-storm (libraryPath = build-storm)

```sh
bash-4.4$ find global-libraries-cache/ -type f -newermt $(date +%Y-%m-%d) | sort -k 8
global-libraries-cache/2079b09ce5b9d359d406cb4c826fc11b39a8bb5c01398d55689261d81ebe1a11-name.txt
global-libraries-cache/2079b09ce5b9d359d406cb4c826fc11b39a8bb5c01398d55689261d81ebe1a11/last_read
global-libraries-cache/2079b09ce5b9d359d406cb4c826fc11b39a8bb5c01398d55689261d81ebe1a11/vars/deleteLibsDir.groovy
global-libraries-cache/2079b09ce5b9d359d406cb4c826fc11b39a8bb5c01398d55689261d81ebe1a11/vars/echoMe.groovy
global-libraries-cache/2079b09ce5b9d359d406cb4c826fc11b39a8bb5c01398d55689261d81ebe1a11/vars/rebaseRepos.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f-name.txt
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/last_read
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/src/lostinberlin/TestClass.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/src/lostinberlin/TestEnum.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/vars/echoMe.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/vars/myPipeline.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/vars/rebuildOnDisconnect.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/vars/retryOnDisconnect.groovy
global-libraries-cache/5a7d69d45c180afdf21fe343b63c3c3d4b7988c0dbc0ac83ff1d0977f7aba33f/vars/updateBuildSchedule.groovy
```